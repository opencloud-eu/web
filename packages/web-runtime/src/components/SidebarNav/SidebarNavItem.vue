<template>
  <li class="oc-sidebar-nav-item oc-pb-xs oc-px-s" :aria-current="active ? 'page' : null">
    <oc-button
      :type="handler ? 'button' : 'router-link'"
      :appearance="active ? 'filled' : 'raw-inverse'"
      color-role="secondaryContainer"
      :class="['oc-sidebar-nav-item-link', 'oc-oc-width-1-1', { active: active }]"
      :data-nav-id="index"
      :data-nav-name="navName"
      :aria-label="
        collapsed ? $gettext('Navigate to %{ pageName } page', { pageName: name }) : undefined
      "
      v-bind="attrs"
    >
      <span class="oc-flex">
        <oc-icon :name="icon" :fill-type="fillType" />
        <span class="oc-ml-m text" :class="{ 'text-invisible': collapsed }" v-text="name" />
      </span>
    </oc-button>
  </li>
</template>
<script lang="ts">
import { FillType } from '@opencloud-eu/design-system/helpers'
import { useRouter } from '@opencloud-eu/web-pkg'
import { computed, defineComponent, PropType, unref } from 'vue'
import { RouteLocationRaw } from 'vue-router'

export default defineComponent({
  props: {
    name: {
      type: String,
      required: true
    },
    index: {
      type: String,
      required: true
    },
    active: {
      type: Boolean,
      required: false,
      default: false
    },
    target: {
      type: [String, Object] as PropType<RouteLocationRaw>,
      required: false,
      default: null
    },
    icon: {
      type: String,
      required: true
    },
    fillType: {
      type: String as PropType<FillType>,
      required: false,
      default: 'fill'
    },
    collapsed: {
      type: Boolean,
      required: false,
      default: false
    },
    handler: {
      type: Function as PropType<() => void>,
      required: false,
      default: null
    }
  },
  setup(props) {
    const router = useRouter()

    const attrs = computed(() => {
      return {
        ...(props.handler && { onClick: props.handler }),
        ...(props.target && { to: props.target })
      }
    })

    const navName = computed(() => {
      if (props.target) {
        return router?.resolve(props.target, unref(router.currentRoute))?.name || 'route.name'
      }
      return props.name
    })

    return { attrs, navName }
  }
})
</script>

<style lang="scss">
.oc-sidebar-nav-item-link {
  position: relative;
  align-items: center !important;
  display: flex !important;
  justify-content: space-between !important;
  padding: var(--oc-space-small) !important;
  border-radius: 5px;
  white-space: nowrap;
  user-select: none;

  .oc-tag {
    background-color: var(--oc-role-tertiary-container);
    color: var(--oc-role-on-tertiary-container);
    svg {
      fill: var(--oc-role-on-tertiary-container);
    }
  }
  .text {
    opacity: 1;
    transition: all 0.3s;
  }
  .text-invisible {
    opacity: 0 !important;
    transition: 0s;
  }

  &.active {
    overflow: hidden;
  }
  &:focus,
  &:hover {
    text-decoration: none !important;
  }

  &:focus:not(.active),
  &:hover:not(.active) {
    background: var(--oc-role-surface-container-highest) !important;
  }

  .oc-icon svg {
    transition: all 0.3s;
  }
}
</style>
