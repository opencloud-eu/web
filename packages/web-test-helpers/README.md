# web-test-helpers

This packages provides utilities for unit testing within the OpenCloud app ecosystem.

## Installation

Depending on your package manager, run one of the following commands:

```
$ npm install @opencloud-eu/web-test-helpers --save-dev

$ pnpm add -D @opencloud-eu/web-test-helpers

$ yarn add @opencloud-eu/web-test-helpers --dev
```

Make sure that you have `vue`, `@vue/test-utils`, and `@opencloud-eu/web-pkg` as dev dependencies of your app since those are peer dependencies of this package.
