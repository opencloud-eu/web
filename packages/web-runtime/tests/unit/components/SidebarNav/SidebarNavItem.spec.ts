import SidebarNavItem from '../../../../src/components/SidebarNav/SidebarNavItem.vue'
import sidebarNavItemFixtures from '../../../__fixtures__/sidebarNavItems'
import { defaultPlugins, mount } from '@opencloud-eu/web-test-helpers'

const exampleNavItem = sidebarNavItemFixtures[0]

const propsData = {
  name: exampleNavItem.name,
  active: false,
  target: exampleNavItem.route.path,
  icon: exampleNavItem.icon,
  index: '5',
  id: '123'
}

describe('OcSidebarNav', () => {
  it('renders navItem without toolTip if expanded', () => {
    const { wrapper } = getWrapper(false)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders navItem with toolTip if collapsed', () => {
    const { wrapper } = getWrapper(true)
    expect(wrapper.html()).toMatchSnapshot()
  })
})

function getWrapper(collapsed: boolean) {
  return {
    wrapper: mount(SidebarNavItem, {
      props: {
        ...propsData,
        collapsed
      },
      global: {
        plugins: [...defaultPlugins()],
        stubs: { 'router-link': true }
      }
    })
  }
}
