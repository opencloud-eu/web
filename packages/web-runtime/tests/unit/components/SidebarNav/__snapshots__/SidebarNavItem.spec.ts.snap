// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OcSidebarNav > renders navItem with toolTip if collapsed 1`] = `
"<li class="oc-sidebar-nav-item oc-pb-xs oc-px-s" id="123">
  <router-link-stub to="/files/list/all" aria-label="Navigate to Personal page" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-secondary-container oc-button-raw-inverse oc-button-secondary-container-raw-inverse oc-sidebar-nav-item-link oc-oc-width-1-1" data-nav-id="5" data-nav-name="route.name"></router-link-stub>
</li>"
`;

exports[`OcSidebarNav > renders navItem without toolTip if expanded 1`] = `
"<li class="oc-sidebar-nav-item oc-pb-xs oc-px-s" id="123">
  <router-link-stub to="/files/list/all" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-secondary-container oc-button-raw-inverse oc-button-secondary-container-raw-inverse oc-sidebar-nav-item-link oc-oc-width-1-1" data-nav-id="5" data-nav-name="route.name"></router-link-stub>
</li>"
`;
